File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.cpp
class Solution {
  public:
    int longestKSubstr(string &s, int k) {
        // code here
        int n=s.size();
        unordered_map<char,int>mp;
        int ans=-1;
        int i=0;
        int j=0;
        while(i<n)
        {
            mp[s[i]]++;
            while(j<i && mp.size()>k)
            {
                --mp[s[j]];
                if(mp[s[j]]==0)
                {
                    mp.erase(s[j]);
                }
                j++;
            }
            if(mp.size()==k)
            {
                ans=max(ans,i-j+1);
            }
            i++;
        }
        return ans;
    }
};

File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.java
class Solution {
    public int longestKSubstr(String s, int k) {
        // code here
        HashMap<Character,Integer>mp=new HashMap<>();
        int n=s.length();
        int j=0;
        int ans=-1;
        int i=0;
        while(i<n)
        {
            mp.put(s.charAt(i),mp.getOrDefault(s.charAt(i),0)+1);
            while(j<i && mp.size()>k)
            {
               mp.put(s.charAt(j),mp.getOrDefault(s.charAt(j),0)-1);
               if(mp.getOrDefault(s.charAt(j),0)==0)
               {
                   mp.remove(s.charAt(j));
                   
               }
               j++;
            }
            if(mp.size()==k)
            {
                ans=Math.max(i-j+1,ans);
            }
            i++;
        }
        return ans;
    }
}